src/Featurizers/Components/StatisticsMetricsEstimator.h

```diff
@@ -0,0 +1,391 @@
+};
+
+template <>
+struct StatsPolicySelector<std::double_t> {
+    using type = StandardStatsTrainingOnlyPolicy<std::double_t>;
+};
+
+} // namespace Details
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       StatisticalMetricsEstimator
+///  \brief         A training-only class that finds the min, max
+///
+template <
+    typename InputT,
+    size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()
+>
+using StatisticalMetricsEstimator                       = TrainingOnlyEstimatorImpl<typename Details::StatsPolicySelector<InputT>::type, MaxNumTrainingItemsV>;
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  BasicStatisticalAnnotationData
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+BasicStatisticalAnnotationData<T>::BasicStatisticalAnnotationData(T min, T max, std::uint64_t count) :
+    Min(std::move(min)),
+    Max(std::move(max)),
+    Count(std::move(count)) {
+        if(Min > Max)
+            throw std::invalid_argument("min is > max");
+
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  StandardStatisticalAnnotaionData
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+StandardStatisticalAnnotaionData<T>::StandardStatisticalAnnotaionData(long double sum, std::double_t average, T min, T max, std::uint64_t count) :
+    BasicStatisticalAnnotationData<T>(min, max, count),
+    Sum(std::move(sum)),
+    Average(std::move(average))
+    {
+        if(Average > BasicStatisticalAnnotationData<T>::Max || Average < BasicStatisticalAnnotationData<T>::Min)
+            throw std::invalid_argument("average is not in the correct range");
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+
+// ----------------------------------------------------------------------
+// |
+// |  update functions
+// |
+// ----------------------------------------------------------------------
+
+template <typename T>
+void update_basic_statistics(T input, T& min, T& max, std::uint64_t& count, bool& first_element_flag) {
+    // first non-null element will set min and max
+    if (first_element_flag) {
+        min = input;
+        max = input;
+        ++count;

```

Comment:
This logic is duplicated with the logic on 298. Right now, that isn't a big deal as it is only a single line. However, this logic will get more complex if another value/additional logic is added. Consider changing this to:

if(first_element_flag) {
    min = input;
    max = input;
    first_element_flag = false;
}
else {
    // Min / max logic here
}

++count;